keps/sig-scheduling/20190312-pod-affinity-and-anti-affinity-support-gt-and-lt-operators.md

```diff
@@ -0,0 +1,170 @@
+ `Pod` affinity/anti-affinity support for `Gt` and `Lt` operators will give users more 
+ control.
+
+ ### Goals
+
+- `Pod` affinity/anti-affinity support for `Gt` and `Lt` operators.
+- `Gt` and `Lt` will have the same status and influence as the original operators (`In`, 
+`NotIn`, `Exists`, `DoesNotExist`).
+
+ ### Non-Goals
+
+- Just add `Gt` and `Lt` operator functions without changing their original definition.
+- Not changing the behavior of other label selectors, such as `Replicases`, `Daemonsets`, etc.
+
+ ## Proposal
+
+ ### User Stories
+
+ As an application developer, I want my application pods to be scheduled onto
+one node that has pod with the "foo" tag and the tag value between "20" and "30".
+
+- if we use the original `In` operator to implement, then we will write affinity 
+like this:
+ ```yaml
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: In
+            values:
+            - 20
+            - 21
+            - 22
+            - 23
+            ...
+            - 28
+            - 29
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+
+ This is not an ideal solution. A promising solution is to provide users with `Gt` and `Lt` 
+ operators, giving users the ability to specify the scope of the tag. Users can achieve 
+ this in this way:
+ ```yaml
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: Gt
+            values:
+            - 20
+          - key: foo
+            operator: Lt
+            values:
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+
+ ### Risks and Mitigations
+
+ Along with this feature, the biggest risk should be performance. This may also be the reason 
+ why the `Gt` and `Lt` operators are not supported when Pod affinity/anti-affinity is 
+ first proposed. In order to eliminate the user's panic, carefully measure the impact of these operators and understand their impact is mandatory. So we will perform performance/benchmark tests on this change and backward compatibility test.
+
+ As part of the mitigations, the good news is that the performance of the entire scheduler has been greatly improved (https://github.com/kubernetes/kubernetes/pull/74041#issuecomment-466191359), and the processing of Pod affinity/anti-affinity has been surprisingly Optimized(https://github.com/kubernetes/kubernetes/pull/67788).
+
+ ## Design Details
+ ### Content
+  We will change the `LabelSelector` of `PodAffinityTerm` to the newly implemented `PodSelector` instead 
+ of `metav1.LabelSelector`:
+
+ ```go
+type PodAffinityTerm struct {
+  LabelSelector *PodSelector
+  ...
+}
+```
+
+ API of `PodSelector` is defined as below:
+
+ ```go
+type PodSelector struct {
```

Comment:
Can we leverage most of node affinity structures here instead of using new ones? I know that the existing ones are hardcoded with prefix `Node` but I believe they're generic enough for example `NodeSelector`, `NodeSelectorRequirements`, `NodeSelectorOperators` have the same fields and types that we could use for pod affinity too.

Reply:
Yes,  @Huang-Wei  also proposed before. The PodSelector was re-proposed to avoid confusion with NodeSelector.  Because from the existing code, many checks and usages of LabelSelector and NodeSelector are inconsistent. I am worried about destroying its original usage habits.

But I think this is a reasonable choice, we can reuse the logic of NodeSelector from a new abstraction.

/cc @bsalamat needs some of your opinion.

Reply:
I agree that the APIs are the same and act in the same way, but given that those have a "Node" prefix, we cannot use them for Pod selection. Otherwise, our API would be very confusing. I like the already proposed solution better that we can create a common `LabelSelectorRequirement` that is used in both Node and Pod selectors.

Reply:
It is a backwards compatible change to change the *name* of the go structs, as those don't actually appear in user config (except for the top level objects like Pod). The field names cannot be changed.

Reply:
If we change the name of go structs, the operators who is using it have to update the code accordingly, right?